allennlp/training/metrics/auc.py

```diff
@@ -0,0 +1,84 @@
+from typing import Optional
+
+from overrides import overrides
+import torch
+from sklearn import metrics
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.training.metrics.metric import Metric
+
+
+@Metric.register("auc")
+class Auc(Metric):
+    """
+    The AUC Metric measures the area under the receiver-operating characteristic
+    (ROC) curve for binary classification problems.
+    """
+    def __init__(self, positive_label=1):
+        super(Auc, self).__init__()
+        self._positive_label = positive_label
+        self._all_predictions = torch.FloatTensor()
+        self._all_gold_labels = torch.LongTensor()
+
+    def __call__(self,
+                 predictions: torch.Tensor,
+                 gold_labels: torch.Tensor,
+                 mask: Optional[torch.Tensor] = None):
+        """
+        Parameters
+        ----------
+        predictions : ``torch.Tensor``, required.
+            A one-dimensional tensor of prediction scores of shape (batch_size).
+        gold_labels : ``torch.Tensor``, required.
+            A one-dimensional label tensor of shape (batch_size), with {1, 0}
+            entries for positive and negative class. If it's not binary,
+            `positive_label` should be passed in the initialization.
+        mask: ``torch.Tensor``, optional (default = None).
+            A one-dimensional label tensor of shape (batch_size).
+        """
+
+        predictions, gold_labels = self.unwrap_to_tensors(predictions, gold_labels)
+
+        # Sanity checks.
+        if gold_labels.dim() != 1:
+            raise ConfigurationError("gold_labels must be one-dimensional, "
+                                     "but found tensor of shape: {}".format(gold_labels.size()))
+        if predictions.dim() != 1:
+            raise ConfigurationError("predictions must be one-dimensional, "
+                                     "but found tensor of shape: {}".format(predictions.size()))
+
+        unique_gold_labels = torch.unique(gold_labels)
+        if unique_gold_labels.numel() > 2:
+            raise ConfigurationError("AUC can be used for binary tasks only. gold_labels has {} unique labels, "
+                                     "expected at maximum 2.".format(unique_gold_labels.numel()))
+
+        gold_labels_is_binary = list(torch.sort(unique_gold_labels)[0].numpy()) == [0, 1]
```

Comment:
This may cause error is some batch contains only one value

Reply:
There is an attempt to fix this in #2775